plugins/module_utils/vmware.py

```diff
@@ -281,3 +287,44 @@ def get_datacenter_detailed(self, name):
 
         """
         return self.get_objs_by_name_or_moid([vim.Datacenter], name)
+
+    def get_datacenter_by_name(self, dc_name, fail_on_missing=False):
```

Comment:
get_datacenter_by_name already exists in vmware_rest...
what is the difference between them?

Reply:
it seems like the all of the `get_` rest methods (or most of them) return the object _identifier_ which is less useful to me. The methods in pyvmomi are returning the actual objects, which is what i need.

Maybe this is a topic better discussed as a group, @machacekondra , @mariolenz. Im not sure I understand the use case for getting an objects identifier instead of the object itself. That seems like its always less useful. 
Additionally, im not sure why we would use both the rest api vs pyvmomi or vice versa if they _can_ return the same infomation

Reply:
> Im not sure I understand the use case for getting an objects identifier instead of the object itself. That seems like its always less useful.

I can't tell you. You see, a lot of the code has been written before my time. _Especially_ a lot of the code in the module utils. If we can simplify things here, let's do it!

Although we have to remember that the module utils (to the best of my knowledge) are considered part of the public API of the collection and breaking changes can only been done in a major release.

> Additionally, im not sure why we would use both the rest api vs pyvmomi or vice versa if they _can_ return the same infomation

A couple of years ago, VMware introduced the new REST API. There were some things possible via the new API, but not via the old (SOAP?) one. On the other hand, most or at least a lot of the stuff that was possible via the old API hasn't been implemented in the new one. Maybe that's the reason? I'll try to dig into this over the weekend, at least for this special case.

ATM I suggest to do as much as possible using pyVmomi because it's easier to install since it is available on PyPI. To the best of my knowledge, the REST SDK is harder to get.

Reply:
I agree, we can refactor the methods to return whatever will be more useful for us in future.

Not sure if module_utils is public API, it's not anywhere published and can't be used directly from playbooks, only internaly by modules, if I am not mistaken. That means if you change the code of the module, you can change the code of module_utils, without affecting users.

Reply:
> Not sure if module_utils is public API, it's not anywhere published and can't be used directly from playbooks, only internaly by modules, if I am not mistaken. That means if you change the code of the module, you can change the code of module_utils, without affecting users.

To the best of my knowledge, playbooks can't use `module_utils`. But _other collections_ can. I'd like to re-use the module utils of this collection in [community.vmware](https://github.com/ansible-collections/community.vmware). So it _is_ part of the public API of your collection.

cc @felixfontein I'm not 100% sure here. I've tried to find some documentation about this, but I wasn't able to. Do you know if we've documented somewhere that module utils are part of a collections API and should follow SemVer? You're usually very well-informed, so maybe you could help here.

Reply:
Generally module_utils are part of the public API of a collection (and thus should follow semver). The inclusion requirements have a section on module_utils that says how to mark them as private to a collection: https://docs.ansible.com/ansible/devel/community/collection_contributors/collection_requirements.html#standards-for-developing-module-and-plugin-utilities

Reply:
cc @mikemorency

Reply:
maybe we should mark them private for a release or two, until theres less changes? As we migrate more stuff from the community collection, I think we will want to address more issues like this one so minor releases would be nice
@mariolenz what do you think about that?

Reply:
> maybe we should mark them private for a release or two, until theres less changes? As we migrate more stuff from the community collection, I think we will want to address more issues like this one so minor releases would be nice @mariolenz what do you think about that?

I think the original question from @bardielle was "get_datacenter_by_name already exists in vmware_rest... what is the difference between them?", wasn't it? Is this answered yet? The answer is different return values, right?

Since this is a _new_ function / method we don't run into a breaking change here. I just wanted to make you aware that you can't refactor the methods to return whatever will be more useful for us in future. At least not once it's been released.

I'm not sure if it's a good idea to mark this as private. It would make it easier for you to change it, but also impossible for other collections to re-use it. I'm really unsure... but since it's a new function / method, maybe you should mark it as private for now if you feel better with this.

> edit: Public -> Private requires a major release as well

Yes, but since it hasn't been released yet I don't see a problem there. Or am I wrong there and it has been released somewhere already?

Reply:
So to summarize:

1. The original question about `get_datacenter_by_name` can be considered resolved. It is a duplicate method that has not been released yet and can be removed with a minor change.
2. The pyvmomi SDK should be prefered if both pyvmomi and the REST sdk provide the same functionality
3. Method names should be descriptive and consistent. Many methods in the community utils will need to be renamed/documented if they are moved to this collection
4. The module utils will be marked as private, temporarily, since theres likely going to be some changes in the next few minor releases related to takeaways 2 and 3. We would like to make it public after that, so community.vmware can leverage the new utils

@mariolenz please let me know if this sounds ok to you and if the PR looks good or needs any other improvements. Thanks to everyone for weighing in on this thread!